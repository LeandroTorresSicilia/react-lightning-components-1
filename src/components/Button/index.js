/* eslint-disable react/require-default-props */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

export default class Button extends Component {
    getBtnClass() {
        const { variant, className } = this.props;

        return classnames('slds-button', {
            'slds-button--neutral': variant === 'neutral',
            'slds-button--brand': variant === 'brand',
            'slds-button--inverse': variant === 'inverse',
            'slds-button--destructive': variant === 'destructive',
            'slds-button--success': variant === 'success',
        }, className);
    }

    render() {
        const {
            label,
            disabled,
            tabIndex,
            onFocus,
            onBlur,
            onClick,
            style,
            children,
            title,
            ariaHaspopup,
        } = this.props;

        return (
            <button className={this.getBtnClass()}
                    disabled={disabled}
                    tabIndex={tabIndex}
                    onFocus={onFocus}
                    onBlur={onBlur}
                    onClick={onClick}
                    title={title}
                    style={style}
                    aria-haspopup={ariaHaspopup} >
                { !children ? label : children }
            </button>
        );
    }
}

Button.propTypes = {
    /** Label for the button. If a children is specified, it will be displayed.
     Otherwise, the component will expect a label which will then displayed */
    label: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.node,
    ]),
    /** Indicates that the element has a popup context menu or sub-level menu. It is
     used for srceen readers */
    ariaHaspopup: PropTypes.bool,
    /** A button predefined style. It must be a string */
    variant: PropTypes.oneOf(['neutral', 'brand', 'inverse', 'destructive', 'success', 'bare']),
    /** This is what will be displayed inside the button. It has precedence over
     the label prop */
    children: PropTypes.node,
    /** Class for custom styles */
    className: PropTypes.string,
    /** Disables the button if set to true */
    disabled: PropTypes.bool,
    /** Tab index */
    tabIndex: PropTypes.number,
    /** Callback function fired when the button is clicked */
    onClick: PropTypes.func,
    /** Callback function fired when the button is focused */
    onFocus: PropTypes.func,
    /** Callback function fired when the button is blurred */
    onBlur: PropTypes.func,
    /** Object with the custom styles. The properties must be
     in camelCase naming convention (e.g. { fontFamily: ‘helvetica’ }) */
    style: PropTypes.object,
    /** This is a description that is showed when a user hover the button */
    title: PropTypes.string,
};

Button.defaultProps = {
    ariaHaspopup: false,
    variant: 'neutral',
    disabled: false,
    onClick: () => {},
    onFocus: () => {},
    onBlur: () => {},
};
