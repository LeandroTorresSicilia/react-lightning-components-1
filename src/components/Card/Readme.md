Card expamples:

    <Card iconName="standard:contact" title="Card Header" footer="Card Footer" actions={<Button label="New" />} >
        Card Body (custom goes here)
    </Card>