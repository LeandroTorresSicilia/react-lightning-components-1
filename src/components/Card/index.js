/* eslint-disable react/require-default-props */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Icon from '../Icon';

export default class Card extends Component {
    getContainerClass() {
        return classnames('slds-card', this.props.className);
    }

    renderIcon() {
        const { iconName } = this.props;

        if (iconName) {
            return (
                <div className="slds-media__figure">
                    <Icon iconName={iconName} size="small" />
                </div>
            );
        }

        return null;
    }

    renderTitle() {
        const { title } = this.props;

        if (title) {
            return (
                <div className="slds-media__body">
                    <h2>
                        <span className="slds-text-heading_small">{title}</span>
                    </h2>
                </div>
            );
        }

        return null;
    }

    renderHeader() {
        const { actions, iconName, title } = this.props;

        if (actions || iconName || title) {
            return (
                <div className="slds-card__header slds-grid">
                    <header className="slds-media slds-media_center slds-has-flexi-truncate">
                        { this.renderIcon() }
                        { this.renderTitle() }
                    </header>
                    { actions ? <div className="slds-no-flex">{actions}</div> : null }
                </div>
            );
        }

        return null;
    }

    render() {
        const {
            actions,
            children,
            footer,
            iconName,
            title,
            style,
        } = this.props;

        if (!actions && !children && !footer && !iconName && !title) {
            return null;
        }

        return (
            <article className={this.getContainerClass()} style={style}>
                { this.renderHeader() }
                { children ? <div className="slds-card__body slds-card__body_inner">{children}</div> : null }
                { footer ? <footer className="slds-card__footer">{footer}</footer> : null }
            </article>
        );
    }
}

Card.propTypes = {
    /** Class for custom styles */
    className: PropTypes.string,
    /** Object with the custom styles. The properties must be in camelCase naming
     convention (e.g. { backgroundColor: green }) */
    style: PropTypes.object,
    /** Actions are components such as button or buttonIcon. Actions are displayed in the header. */
    actions: PropTypes.node,
    /** The body of the component. In markup, this is everything in the body of the card */
    children: PropTypes.node,
    /** The footer can include text or another component */
    footer: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
    /** The Lightning Design System name of the icon. Names are written in the
     format '\utility:down\' where 'utility' is the category, and 'down' is the
     specific icon to be displayed. The icon is displayed in the header to the left of the title */
    iconName: PropTypes.string,
    /** The title can include text or another component, and is displayed in the header */
    title: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
};
