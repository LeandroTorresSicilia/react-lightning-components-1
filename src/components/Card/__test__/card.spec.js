import React from 'react';
import renderer from 'react-test-renderer';
import Card from './../index';
import Button from './../../Button';

describe('<Card/>', () => {
    it('should render the right markup for the card component', () => {
        const component = renderer.create(
            <Card iconName="standard:contact" title="Card Header" footer="Card Footer" actions={<Button label="New" />} >
                Card Body (custom goes here)
            </Card>,
        );

        expect(component).toMatchSnapshot();
    });

    it('should render nothing when the card is created without any prop', () => {
        const component = renderer.create(<Card />);

        expect(component).toMatchSnapshot();
    });

    it('should render the right markup for the card body when only children is passed', () => {
        const component = renderer.create(<Card>Card Body</Card>);

        expect(component).toMatchSnapshot();
    });

    it('should render the right markup for the card footer when only footer is passed', () => {
        const component = renderer.create(<Card footer="Card footer" />);

        expect(component).toMatchSnapshot();
    });

    it('should render the right markup for the card header when only iconName is passed', () => {
        const component = renderer.create(<Card iconName="standard:contact" />);

        expect(component).toMatchSnapshot();
    });

    it('should render the right markup for the card header when only title is passed', () => {
        const component = renderer.create(<Card title="Card title" />);

        expect(component).toMatchSnapshot();
    });

    it('should render the right markup for the card header when only actions is passed', () => {
        const component = renderer.create(<Card actions={<Button label="New" />} />);

        expect(component).toMatchSnapshot();
    });
});
