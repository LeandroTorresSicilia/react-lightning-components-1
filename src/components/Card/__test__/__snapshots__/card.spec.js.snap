// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card/> should render nothing when the card is created without any prop 1`] = `null`;

exports[`<Card/> should render the right markup for the card body when only children is passed 1`] = `
<article
  className="slds-card"
  style={undefined}
>
  <div
    className="slds-card__body slds-card__body_inner"
  >
    Card Body
  </div>
</article>
`;

exports[`<Card/> should render the right markup for the card component 1`] = `
<article
  className="slds-card"
  style={undefined}
>
  <div
    className="slds-card__header slds-grid"
  >
    <header
      className="slds-media slds-media_center slds-has-flexi-truncate"
    >
      <div
        className="slds-media__figure"
      >
        <span
          className="slds-icon_container slds-icon-standard-contact"
          style={undefined}
          title={undefined}
        >
          <svg
            aria-hidden="true"
            className="slds-icon_small slds-icon"
            style={undefined}
          >
            <use
              xlinkHref="/icons/standard-sprite/svg/symbols.svg#contact"
              xmlnsXlink="http://www.w3.org/1999/xlink"
            />
          </svg>
          <span
            className="slds-assistive-text"
          />
        </span>
      </div>
      <div
        className="slds-media__body"
      >
        <h2>
          <span
            className="slds-text-heading_small"
          >
            Card Header
          </span>
        </h2>
      </div>
    </header>
    <div
      className="slds-no-flex"
    >
      <button
        aria-haspopup={false}
        className="slds-button slds-button--neutral"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        style={undefined}
        tabIndex={undefined}
        title={undefined}
      >
        New
      </button>
    </div>
  </div>
  <div
    className="slds-card__body slds-card__body_inner"
  >
    Card Body (custom goes here)
  </div>
  <footer
    className="slds-card__footer"
  >
    Card Footer
  </footer>
</article>
`;

exports[`<Card/> should render the right markup for the card footer when only footer is passed 1`] = `
<article
  className="slds-card"
  style={undefined}
>
  <div
    className="slds-card__body slds-card__body_inner"
  />
  <footer
    className="slds-card__footer"
  >
    Card footer
  </footer>
</article>
`;

exports[`<Card/> should render the right markup for the card header when only actions is passed 1`] = `
<article
  className="slds-card"
  style={undefined}
>
  <div
    className="slds-card__header slds-grid"
  >
    <header
      className="slds-media slds-media_center slds-has-flexi-truncate"
    />
    <div
      className="slds-no-flex"
    >
      <button
        aria-haspopup={false}
        className="slds-button slds-button--neutral"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        style={undefined}
        tabIndex={undefined}
        title={undefined}
      >
        New
      </button>
    </div>
  </div>
  <div
    className="slds-card__body slds-card__body_inner"
  />
</article>
`;

exports[`<Card/> should render the right markup for the card header when only iconName is passed 1`] = `
<article
  className="slds-card"
  style={undefined}
>
  <div
    className="slds-card__header slds-grid"
  >
    <header
      className="slds-media slds-media_center slds-has-flexi-truncate"
    >
      <div
        className="slds-media__figure"
      >
        <span
          className="slds-icon_container slds-icon-standard-contact"
          style={undefined}
          title={undefined}
        >
          <svg
            aria-hidden="true"
            className="slds-icon_small slds-icon"
            style={undefined}
          >
            <use
              xlinkHref="/icons/standard-sprite/svg/symbols.svg#contact"
              xmlnsXlink="http://www.w3.org/1999/xlink"
            />
          </svg>
          <span
            className="slds-assistive-text"
          />
        </span>
      </div>
    </header>
  </div>
  <div
    className="slds-card__body slds-card__body_inner"
  />
</article>
`;

exports[`<Card/> should render the right markup for the card header when only title is passed 1`] = `
<article
  className="slds-card"
  style={undefined}
>
  <div
    className="slds-card__header slds-grid"
  >
    <header
      className="slds-media slds-media_center slds-has-flexi-truncate"
    >
      <div
        className="slds-media__body"
      >
        <h2>
          <span
            className="slds-text-heading_small"
          >
            Card title
          </span>
        </h2>
      </div>
    </header>
  </div>
  <div
    className="slds-card__body slds-card__body_inner"
  />
</article>
`;
